public/app/features/correlations/Forms/TransformationsEditor.tsx

```diff
@@ -0,0 +1,281 @@
+import { Stack } from '@grafana/experimental';
+import {
+  Button,
+  Field,
+  FieldArray,
+  Icon,
+  IconButton,
+  Input,
+  InputControl,
+  Label,
+  Select,
+  Tooltip,
+  useStyles2,
+} from '@grafana/ui';
+
+type Props = { readOnly: boolean };
+
+const getStyles = (theme: GrafanaTheme2) => ({
+  heading: css`
+    font-size: ${theme.typography.h5.fontSize};
+    font-weight: ${theme.typography.fontWeightRegular};
+  `,
+  // set fixed position from the top instead of centring as the container
+  // may get bigger when the for is invalid
+  removeButton: css`
+    margin-top: 25px;
+  `,
+});
+
+export const TransformationsEditor = (props: Props) => {
+  const { control, formState, register, setValue, watch, getValues } = useFormContext();
+  const { readOnly } = props;
+  const [keptVals, setKeptVals] = useState<{ [key: string]: { expression: string; mapValue: string } }>({});
+
+  const styles = useStyles2(getStyles);
+
+  const transformOptions = getTransformOptions();
+
+  return (
+    <FieldArray name="config.transformations" control={control}>
+      {({ fields, append, remove }) => (
+        <>
+          <Stack direction="column" alignItems="flex-start">
+            <div className={styles.heading}>Transformations</div>
+            {fields.length === 0 && <div> No transformations defined.</div>}
+            {fields.length > 0 && (
+              <div>
+                {fields.map((field, index) => {
+                  return (
+                    <Stack direction="row" key={field.id} alignItems="top">
+                      <Field
+                        label={
+                          <Stack gap={0.5}>
+                            <Label>Type</Label>
+                            <Tooltip
+                              content={
+                                <div>
+                                  <p>The type of transformation that will be applied to the source data.</p>
+                                </div>
+                              }
+                            >
+                              <Icon name="info-circle" size="sm" />
+                            </Tooltip>
+                          </Stack>
+                        }
+                        invalid={!!formState.errors?.config?.transformations?.[index]?.type}
+                        error={formState.errors?.config?.transformations?.[index]?.type?.message}
+                        validationMessageHorizontalOverflow={true}
+                      >
+                        <InputControl
+                          render={({ field: { onChange, ref, ...field } }) => (
+                            <Select
+                              {...field}
+                              onChange={(value) => {
+                                if (!readOnly) {
+                                  const currentValues = getValues().config.transformations[index];
+                                  setKeptVals({
```

Comment:
Keeping some values is for the scenario where people might click around the editor and not realize they're quasi losing data. If someone has a regex with values, and selects logfmt (which wipes those values out) and then goes back to regex, i would like the previously entered values to show back up. Those values are always captured before they change due to a transformation type change.